egs2/interspeech2024_dsu_challenge/asr2/local/score.sh

```diff
@@ -0,0 +1,123 @@
+        } END {
+            for (inf_conf in dset) {
+                n=split(dset[inf_conf], lst, "#");
+                if (n==4) {
+                    printf("%s/EN_LibriSpeech %.2f\n", inf_conf, 100*nerr_sum[inf_conf]/nchar_sum[inf_conf])
+                }
+            }
+        }')
+    # echo "${final_ls_result}"
+
+else
+    echo "Error: no librispeech eval set results found in ${asr_exp}. Exit." && exit 1;
+fi
+
+# ML-SUPERB eval sets
+if ls "${asr_exp}"/*/test_1h/score_cer/result.txt &> /dev/null; then
+    ms_result=$(grep -H -e Sum "${asr_exp}"/*/test_1h/score_cer/result.txt | \
+        grep -v -e Avg | \
+        sed -e "s#${asr_exp}/\([^/]*/[^/]*\)/score_cer/result.txt:#|\1#g" | \
+        sed -e 's#Sum##g' | tr '|' ' ' | tr -s ' ' '|')
+    final_ms_result=$(echo "${ms_result}" | awk -F"|" '{
+            n=split($2, lst, "/");
+            nutt_sum[lst[1]]=$3; nchar_sum[lst[1]]=$4; nerr_sum[lst[1]]=$9;
+            dset[lst[1]]=lst[2];
+        } END {
+            for (inf_conf in dset) {
+                if (dset[inf_conf] == "test_1h") {
+                    printf("%s/ML_SUPERB %.2f\n", inf_conf, 100*nerr_sum[inf_conf]/nchar_sum[inf_conf])
+                }
+            }
+        }')
+    # echo "${final_ms_result}"
+
+else
+    echo "Error: no ml_superb eval set results found in ${asr_exp}. Exit." && exit 1;
+fi
+
+if [[ -n ${final_ls_result} ]] && [[ -n ${final_ms_result} ]]; then
+
+    bitrate=$(python3 pyscripts/utils/calculate_bitrate.py \
+        --vocab \
+        --tokens \
+        --reference_len \
+        )
+
+    echo "## ${asr_exp}"
+    echo "${ls_result}"
+    echo "${ms_result}"
+    echo
+
+    all_results=$(echo "${final_ls_result}"; echo "${final_ms_result}")
+    echo "${all_results}"
+    echo
+
+    echo "${all_results}" | awk '{
+        n=split($1, lst, "/");
+        if (lst[2]=="EN_LibriSpeech") {results[lst[1]]["en"]=$2;}
+        else if (lst[2]=="ML_SUPERB") {results[lst[1]]["ml"]=$2;}
+        } END {
+            for (inf_conf in results) {
+                if (length(results[inf_conf])==2) {
+                    print(inf_conf, results[inf_conf]["en"], results[inf_conf]["ml"])
+                }
+            }
+        }' | while read line; do
+            output_file=$(echo "${line}" | cut -d" " -f1)"_result.json"
+            echo "Writing submission json file in ${asr_exp}/${output_file}"
+            echo
+
+            cat <<EOF > ${asr_exp}/${output_file}
+{
+    "config": {
+        "model_dtype": "torch.float16",
+        "model_name": "YOUR_MODEL_NAME"
+    },
+    "results": {
+        "asr_eval1": {
+            "EN_LibriSpeech": $(echo "${line}" | cut -d" " -f2)
+        },
+        "asr_eval2": {
+            "ML_SUPERB": $(echo "${line}" | cut -d" " -f3)
+        },
+        "asr_eval3": {
+            "Bitrate": XX.XX
+        }
+    }
+}
+EOF
+
+    done
+
+    echo "Please compute the Bitrate information manually by following the README.md."
```

Comment:
Is it difficult to integrate it?
If it takes time, we can leave it for now, but it is better to run the bitrate calculation at the same time.

Reply:
Sorry, I tried to do it yesterday. It was not that straightforward to integrate it.
Maybe I was wrong. Let me do it now.

Reply:
thanks!
That is great.
But if it takes time, we can leave it for now.

Reply:
I double-check it. Technically, it is almost done. As written in the README, the last python block can get the bitrate number as a variable. 
```bash
$ token_suffix="wavlm_large_21_km2000"
$ bpe_folder="data/token_list/src_bpe_unigram3000_rm_wavlm_large_21_km2000"
$ for dset in dev_clean dev_other test_clean test_other test_1h; do
$   result_dir="bitrate/${dset}"

$   paste \
$     <(<dump/raw/${dset}/text.rm.${token_suffix} cut -d" " -f1) \
$     <(<dump/raw/${dset}/text.rm.${token_suffix} spm_encode --model=${bpe_folder}/bpe.model --output_format=id) \
$     > dump/raw/${dset}/token_int.rm.${token_suffix}

$   python pyscripts/utils/convert_token2json.py \
$     --vocab data/token_list/src_bpe_unigram3000_rm_${token_suffix}/tokens.txt \
$     --token dump/raw/${dset}/token_int.rm.${token_suffix} \
$     --ref_scp data/${dset}/wav.scp \
$     --result_dir "${result_dir}"

$   python pyscripts/utils/calculate_bitrate.py \
$     --vocab ${result_dir}/vocab.json \
$     --tokens ${result_dir}/tokens.json \
$     --reference_len ${result_dir}/ref_len.scp \
$     --bitrate_details ${result_dir}/details.txt
$ done

$ python - <<EOF
import numpy as np
bitrates=[]
result_dir="${result_dir}/.."
for dset in ["dev_clean", "dev_other", "test_clean", "test_other", "test_1h"]:
    with open(f"{result_dir}/{dset}/details.txt", "r") as f:
        for line in f.readlines():
            lst = line.strip().split()
            bitrates.append(float(lst[1]))
print(np.mean(bitrates))

EOF
```

However, the only issue is the bpe folder name defined in asr2.sh: ```src_bpedir="${token_listdir}/src_bpe_${src_bpemode}${src_nbpe}_${src_case}_${src_lang}"```.
This is not directly passed to `local/score.sh`. If I add this argument, other recipes may have error.

Reply:
I see.
This is difficult...
@ftshijt, do you have some ideas?
If not, maybe we should give up...